### PR TITLE
[AF-2768] Add time series component

### PR DESCRIPTION
**Pull request by manaswinidas**

### CODE REVIEW

File: packages/time-series-chart/package.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      ">0.2%",
+      "not dead",
+      "not op_mini all"
+    ],
+    "development": [
+      "last 1 chrome version",
+      "last 1 firefox version",
+      "last 1 safari version"
+    ]
+  },
```
Comment (jesuino) [inline packages/time-series-chart/package.json:49]: Why were these lines added? Could it be removed?

File: packages/time-series-chart/src/Data.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  };
+  xaxis: {
+    categories: Array<string | number>;
+  };
+}
+
+export interface typeseries {
+  name: string;
+  data: Array<number>;
+}
```
Comment (jesuino) [inline packages/time-series-chart/src/Data.tsx:13]: I would use same type name standard used with Java type and remove the "type" prefix. It could be: Options and Series or ApexOptions or ApexSeries.

File: packages/time-series-chart/src/LineChart.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import  Chart from "react-apexcharts";
+
+export interface ChartProps {
+  width?: number;
+  height?: number;
+  options: typeoptions;
+  series: typeseries;
+}
+
+export interface TreemapState {}
```
Comment (jesuino) [inline packages/time-series-chart/src/LineChart.tsx:12]: I think it is lost here

File: packages/time-series-chart/src/react-app-env.d.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
Comment (jesuino) [inline packages/time-series-chart/src/react-app-env.d.ts:1]: do we need this file?

File: packages/time-series-chart/src/serviceWorker.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
+// This optional code is used to register a service worker.
```
Comment (jesuino) [inline packages/time-series-chart/src/serviceWorker.ts:1]: do we need this file?

File: packages/time-series-chart/src/setupTests.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+// jest-dom adds custom jest matchers for asserting on DOM nodes.
```
Comment (jesuino) [inline packages/time-series-chart/src/setupTests.ts:1]: do we need this file?

**Review by jesuino — CHANGES_REQUESTED** [review]
Hello @manaswinidas 

Thanks for your component, it looks great and works as expected, I made some comments for improvements.

I would also suggest, if possible, to support multiple series. Here's a sample of a Prometheus dataset result:

image
What I wanted is to have a column with series values and another column for category, here's an example:

CAT - SERIES - VALUE
C1   - S1          - 0
C1   - S2          - 1
C2   - S1          - 2
C2   - S2          - 3

In the chart we would have one line for S1 and other for S2

Would it be possible?

Thanks!

File: packages/time-series-chart/src/Chart.tsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+// Default Values
+const DEFAULT_WIDTH = 400;
+const DEFAULT_HEIGHT = 200;
+const NOT_ENOUGH_COLUMNS_MSG =
+  "Time series component expects 2 columns: category(LABEL or TEXT or NUMBER or DATE) and series(NUMBER).";
+const FIRST_COLUMN_INVALID_MSG = "Wrong type for first column, it should be either LABEL, TEXT, NUMBER or DATE.";
+const SECOND_COLUMN_INVALID_MSG = "Wrong type for second column, it should be NUMBER.";
+
+enum Params {
+  WIDTH = "width",
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:17]: Could we make it fill the whole body (width 100%) by default?

Take as example Heatmap:

https://github.com/kiegroup/appformer/tree/master/dashbuilder/dashbuilder-shared/dashbuilder-js/packages/process-heatmap-component

It fill the whole page by default and the width/height setting can be controlled on Dashbuilder side, where we control the parent width and height

File: packages/time-series-chart/src/Chart.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  if (ds.columns[1].type !== ColumnType.NUMBER) {
+    return SECOND_COLUMN_INVALID_MSG;
+  }
+};
+
+interface Props {
+  controller: ComponentController;
+}
+
+function getseries(dataset: any): any {
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:67]: Please use the actual type on return, not any or we lose strong typing brought by typescript:

```
function getseries(dataset: DataSet): SingleSeries[] {
....
```

File: packages/time-series-chart/src/Chart.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  var arrayseries: Array<SingleSeries> = [];
+  for( var i=1; i<dataset.columns.length; i++){
+  var newseries: SingleSeries = { name: dataset.columns[i].name, data: dataset.data.map((d: any) => d[i]) };
+  arrayseries.push(newseries);
+  }
+  return arrayseries;
+}
+
+function getoptions(dataset: any): any {
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:76]: Please use types for the parameter and return, avoid any as possible, here's a suggested refactor:
```
function getoptions(dataset: DataSet): Options {
   return {
    chart: { id: dataset.columns[0].name },
    xaxis: { categories: dataset.data.map((d: any) => d[1]) }
  };
}
```

File: packages/time-series-chart/src/Chart.tsx [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+  var newoptions: Options = {
+    chart: { id: dataset.columns[0].name },
+    xaxis: { categories: dataset.data.map((d: any) => d[1]) }
+  };
+  return newoptions;
+}
+
+export function Chart(props: Props) {
+  const [chartProps, setChartProps] = useState<ChartProps>({
+    width: DEFAULT_WIDTH,
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:86]: If we make it occupy the whole body by default then we can avoid this additional configuration

**Review by jesuino — CHANGES_REQUESTED** [review]
Hello @manaswinidas 

Thanks for the updates, it looks cleaner without the additional files.

I made some suggestions. In summary we should avoid setting width/height on the component level because we have this option on DB level. Make the component always full the screen and be responsible would be the best for us. We had trouble with it with heatmaps as well, but the result is a diagram that can adapt easily to any page where it is placed.

The other point is about "any" usage. We should avoid it and use types - any should be used only in some specific cases.

THanks!

File: packages/time-series-chart/package.json [hunk 11]
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "monitoring-donuts",
```
Comment (jesuino) [inline packages/time-series-chart/package.json:2]: Could you please correct the name?

File: packages/time-series-chart/static/manifest.json [hunk 12]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "time series",
+  "icon": "fa fa-bell-o",
+  "parameters": [
+    {
+      "name": "flag",
```
Comment (jesuino) [inline packages/time-series-chart/static/manifest.json:6]: I would name this parameter something else, like "Transpose" and make it boolean. When transpose is on then users must provide a column for the series instead using individual columns for the series.

File: packages/time-series-chart/static/index.html [hunk 13]
```diff
@@ -0,0 +1,30 @@
+<!--
+  ~ Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
Comment (jesuino) [inline packages/time-series-chart/static/index.html:2]: Could you please update to "2021" and also add license headers to the other files that we can add comment (see my others comments)

File: packages/time-series-chart/src/SampleData.tsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+export const options = {
```
Comment (jesuino) [inline packages/time-series-chart/src/SampleData.tsx:1]: Add license header

File: packages/time-series-chart/src/LineChart.tsx [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
```
Comment (jesuino) [inline packages/time-series-chart/src/LineChart.tsx:1]: add license header

File: packages/time-series-chart/src/Data.tsx [hunk 16]
```diff
@@ -0,0 +1,13 @@
+export interface Options {
```
Comment (jesuino) [inline packages/time-series-chart/src/Data.tsx:1]: add license header

File: packages/time-series-chart/src/Chart.tsx [hunk 17]
```diff
@@ -0,0 +1,203 @@
+import * as React from "react";
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:1]: add license header

File: packages/time-series-chart/src/Chart.tsx [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+  }
+};
+
+const validateDataSet = (ds: DataSet, flag: number): string | undefined => {
+  if (flag === 0) {
+    if (ds.columns.length < 2) {
+      return NOT_ENOUGH_COLUMNS_MSG;
+    }
+    if (
+      ds.columns[0].type !== ColumnType.LABEL &&
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:55]: Notice that this condition will never be evaluated to true because you covered all column types, a column can only be TEXT, LABEL, NUMBER or DATE.

File: packages/time-series-chart/src/Chart.tsx [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+  if (!params.get(Params.FLAG)) {
+    return "Flag is required.";
+  }
+  if (!params.get(Params.CHARTNAME)) {
+    return "Chart name is required.";
+  }
+};
+
+const validateDataSet = (ds: DataSet, flag: number): string | undefined => {
+  if (flag === 0) {
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:50]: I would suggest you to rename flag to "transpose" and use it as a boolean variable

File: packages/time-series-chart/src/Chart.tsx [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+      ds.columns[0].type !== ColumnType.DATE
+    ) {
+      return FIRST_COLUMN_INVALID_MSG;
+    }
+    if (ds.columns[1].type !== ColumnType.NUMBER) {
+      return SECOND_COLUMN_INVALID_MSG;
+    }
+  }
+  if (flag === 1) {
+    if (ds.columns.length < 3) {
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:67]: For clarity I also suggest you to extract the validation for transposed dataset and not transposed in different methods:
```
if(transposed) {
     return validateTransposedDataSet(dataSet);
}
else {
     // do the validation for not transposed datasets
}
```

File: packages/time-series-chart/src/Chart.tsx [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+    }
+    if (
+      ds.columns[0].type !== ColumnType.LABEL &&
+      ds.columns[0].type !== ColumnType.TEXT &&
+      ds.columns[0].type !== ColumnType.NUMBER &&
+      ds.columns[0].type !== ColumnType.DATE
+    ) {
+      return FIRST_COLUMN_INVALID_MSG;
+    }
+    if (ds.columns[1].type !== ColumnType.NUMBER) {
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:62]: Notice that now we support multiple series then you need to validate Columns[1...n] and not only the first column!

File: packages/time-series-chart/src/Chart.tsx [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+function getseries(dataset: DataSet, flag: number): Array<SingleSeries> {
+  var arrayseries: Array<SingleSeries> = [];
+  var newseries: SingleSeries = { name: "", data: [] };
+  if (flag === 0) {
+    for (var i = 1; i < dataset.columns.length; i++) {
+      newseries = { name: dataset.columns[i].name, data: dataset.data.map((d: Array<string | number>) => d[i]) };
+      arrayseries.push(newseries);
+    }
+  }
+  if (flag === 1) {
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:100]: I would separate in a method the series extraction for transposed data set.
Separate the methods make it easier to test.

**Review by jesuino — CHANGES_REQUESTED** [review]
Hello @manaswinidas 

Thanks for your continued great work on this, we are getting there! Your implementation for transposed dataset looks great, we just need some improvements.

Since we are about to wrap up this work, I am asking you to add license headers because soon we will move this component to the official repository and make it an internal component.


I noticed that the component bottom is not being displayed anymore, could you please check it? In any case, it looks great in full screen.

image

Finally there are quite a few problems with linting. Kindly run "yarn run build:prod" on your project and you will see all warnings that should be fixed.

Thanks and let me know if you have any question.

File: packages/time-series-chart/src/Chart.tsx (hunk 18 above)
Comment (manaswinidas) [reply packages/time-series-chart/src/Chart.tsx:55]: `if (!Object.values(ColumnType).includes(ds.columns[0].type))` is this any better?

File: packages/time-series-chart/src/Chart.tsx [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+    arrayseries = getSeriesforNonTransposedDataset(dataset);
+  }
+  return arrayseries;
+}
+
+function getSeriesforNonTransposedDataset(dataset: DataSet): SingleSeries[] {
+  const arrayseries: SingleSeries[] = [];
+  let newseries: SingleSeries = { name: "", data: [] };
+  for (let i = 1; i < dataset.columns.length; i++) {
+    newseries = { name: dataset.columns[i].name, data: dataset.data.map((d: Array<string | number>) => d[i]) };
```
Comment (jesuino) [inline packages/time-series-chart/src/Chart.tsx:113]: Would it be possible to use `dataset.columns[i].settings.columnName` instead `columns[i].name` - this way the component will have the column that user configured in the UI instead the column id.

**Review by jesuino — CHANGES_REQUESTED** [review]
Hello Das,

It is looking good and working well on Dashbuilder. I have a few more minor suggestions for improvements.

I noticed that sizing issue is related to Apex chart internal resizing. We may revisit it in future, but for now we will have to adjust the component size to make sure that the full chart is displayed.

THanks!

**Comment by jesuino** [comment]
Hello @manaswinidas 

Thanks for all the fixes, it looks good to me now!

A last thing to ask is could you please bring back scroll? I made a brief research on Apex chart responsiveness and looks like it tries to keep aspect ratio when resizing, I think hiding the bottom may confuse users, hence your previous suggestion is great. Let's make the body `overflow: auto` as you suggested.

Thanks!

Reference about aspect ratio: https://apexcharts.com/docs/options/chart/height/

**Comment by jesuino** [comment]
Thank you @manaswinidas I am merging your PR now.
